cpp/doc/TESTING.md

```diff
@@ -0,0 +1,105 @@
+# Unit Testing in libcuspatial
+
+Unit tests in libcuspatial are written using
+[Google Test](https://github.com/google/googletest/blob/master/docs/primer.md).
+
+## Best Practices: What Should We Test?
+
+In general we should test to make sure all code paths are covered. This is not always easy or
+possible. But generally this means we test all supported combinations of algorithms and data types,
+and the main iterator and container types supported by algorithms.  Here are some other guidelines.
+
+ * In general empty input is not an error in libcuspatial. Typically empty input results in empty
+   output. Tests should verify this.
+
+ * Most algorithms should have one or more tests exercising inputs with a large enough number of
+   rows to require launching multiple thread blocks, especially when values are ultimately
+   communicated between blocks (e.g. reductions). This is especially important for custom kernels
+   but also applies to Thrust and CUB algorithm calls with lambdas / functors.
+
+ * Test exceptional cases. For example, anything that causes the function to `throw`.
+
+ * Test boundary cases. For example points that fall exactly on lines or boundaries.
+
+## Header-only and Column-based API tests
+
+libcuspatial currently has two C++ APIs: the column-based API uses libcudf data structures as 
+input and output. These tests live in `cpp/tests/` and can use libcudf features for constructing
+columns and tables. The header-only API does not depend on libcudf at all and so tests of these
+APIs should not include any libcudf headers. These tests currently live in `cpp/tests/experimental`.
```

Comment:
What's unstated here is that we generally leave all "core" test cases (test cases that's specific to computation) to the experimental API, and "peripheral" test cases (test cases that's specific to column API on type handling, input validation etc.) to the column API.

The compile time requirements on the experimental APIs are generally not tested. Because they don't pass compile.

Reply:
Done.